src/types/list.rs

```diff
@@ -462,29 +483,119 @@ impl<'py> PyListMethods<'py> for Bound<'py, PyList> {
     }
 }
 
+// New types for type checking when using BoundListIterator associated methods, like
+// BoundListIterator::next_unchecked.
+struct Index(usize);
+struct Length(usize);
+
 /// Used by `PyList::iter()`.
 pub struct BoundListIterator<'py> {
     list: Bound<'py, PyList>,
-    index: usize,
-    length: usize,
+    index: Index,
+    length: Length,
+}
+
+/// Helper to manage mutable borrows below
+macro_rules! split_borrow {
+    ($instance:expr, $index:ident, $length:ident, $list:ident) => {
+        let Self {
+            ref mut $index,
+            ref mut $length,
+            ref $list,
+        } = $instance;
+    };
 }
```

Comment:
rather than just using this to split the borrow, maybe we should just have a macro that wraps
```rust
        crate::sync::with_critical_section(list, || {
            ...
        })
``` as well?

Reply:
I tried to do this this afternoon and got stuck. I'm pretty novice at writing macros - do you know of any similar examples I could look at somewhere?

Here's what I tried, but this doesn't compile if I used it in e.g. the `fold` implementation: https://gist.github.com/ngoldbaum/13ac11629f042ae0bee84559e4e8bb31

Reply:
@ngoldbaum was the error about identifiers not being found? If yes it's because Rust macros are hygienic, all referenced identifiers in a macro argument must be defined outside of the macro. What you can do is something like this:

```rust
/// Helper to manage mutable borrows below
macro_rules! split_borrow {
    ($instance:expr, $index:ident, $length:ident, $list:ident) => {
        let Self {
            ref mut $index,
            ref mut $length,
            ref $list,
        } = $instance;
    };
}

macro_rules! op_with_critical_section {
    ($instance:expr, $fn:expr) => {{
        split_borrow!($instance, index, length, list);
        crate::sync::with_critical_section(list, || $fn(index, length, list))
    }};
}

impl<'py> Iterator for BoundListIterator<'py> {
    type Item = Bound<'py, PyAny>;

    fn fold<B, F>(mut self, init: B, mut f: F) -> B
    where
        Self: Sized,
        F: FnMut(B, Self::Item) -> B,
    {
        op_with_critical_section!(self, |index: &mut Index, length: &mut Length, list| {
            let mut accum = init;
            while let Some(x) = unsafe { Self::next_unchecked(&mut *index, &mut *length, list) } {
                accum = f(accum, x);
            }
            accum
        })
    }
}
```

Though at this point I'm not sure the macro redirections makes it more readable anymore...